src/kms/DeterministicCrypto.ts

```diff
@@ -0,0 +1,227 @@
+import Future from "futurejs";
+import {TenantSecurityErrorCode, TenantSecurityException} from "../TenantSecurityException";
+import {TenantSecurityExceptionUtils} from "../TenantSecurityExceptionUtils";
+import {TscException} from "../TscException";
+import * as miscreant from "miscreant";
+import {
+    BatchResult,
+    DeterministicEncryptedField,
+    DeterministicEncryptedFieldCollection,
+    DeterministicPlaintextField,
+    DeterministicPlaintextFieldCollection,
+    mapBatchOperationToResult,
+} from "../Util";
+import {DETERMINISTIC_HEADER_PADDING} from "./Constants";
+import {DerivedKey, DeriveKeyResponse, getDerivedKeys} from "./KmsApi";
+
+const cryptoProvider = new miscreant.PolyfillCryptoProvider();
+
+interface DeterministicEncryptedFieldParts {
+    tenantSecretId: number;
+    encryptedBytes: Buffer;
+}
+
+export type BatchDeterministicEncryptResult = TenantSecurityException | DeterministicEncryptedField;
+export type BatchDeterministicDecryptResult = TenantSecurityException | DeterministicPlaintextField;
+export type BatchDeterministicSearchResult = TenantSecurityException | DeterministicEncryptedField[];
+
+/**
+ * Deterministically encrypt the provided field with the current key from `derivedKeys`.
+ */
+export const encryptField = (
+    field: DeterministicPlaintextField,
+    derivedKeys: DerivedKey[] | undefined
+): Future<TenantSecurityException, DeterministicEncryptedField> => {
+    const current = derivedKeys?.find((key) => key.current);
+    if (current === undefined) {
+        return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_ENCRYPT_FAILED, "Failed deterministic encryption."));
+    }
+    const dekBytes = Buffer.from(current.derivedKey, "base64");
+    const encryptedFuture = Future.gather2(generateEncryptedFieldHeader(current.tenantSecretId), encryptBytes(field.plaintextField, dekBytes)).map(
+        ([header, encryptedDoc]) => Buffer.concat([header, encryptedDoc])
+    );
+    return encryptedFuture.map((encrypted) => ({
+        encryptedField: encrypted,
+        derivationPath: field.derivationPath,
+        secretPath: field.secretPath,
+    }));
+};
+
+/**
+ * Encode the tenant secret ID as 4 bytes, then attach two bytes of 0s. Fails if the tenant secret ID can't fit into 4 bytes.
+ */
+export const generateEncryptedFieldHeader = (tenantSecretId: number): Future<TenantSecurityException, Buffer> => {
+    if (tenantSecretId < 0 || tenantSecretId > Math.pow(2, 32) - 1) {
```

Comment:
We should compute the max value and put it in a constant so we're not continually doing that math.

That value is 4294967295.

Reply:
I had meant to and forgot. Added it now